compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts

```diff
@@ -0,0 +1,517 @@
+        lvalue.identifier.declarationId,
+      );
+      if (value.kind === 'PropertyLoad') {
+        const property = getProperty(
+          value.object,
+          value.property,
+          temporaries,
+          properties,
+        );
+        if (!usedOutside) {
+          properties.set(lvalue.identifier, property);
+        }
+        propertyLoads.push({
+          instr: instr as TInstruction<PropertyLoad>,
+          property,
+        });
+      } else if (value.kind === 'LoadLocal') {
+        if (
+          lvalue.identifier.name == null &&
+          value.place.identifier.name !== null &&
+          !usedOutside
+        ) {
+          temporaries.set(lvalue.identifier, value.place.identifier);
+        }
+      }
+    }
+    propertyLoadInfo.set(block.id, propertyLoads);
+  }
+  return {temporaries, properties, propertyLoadInfo};
+}
+
+function collectNodes(
+  fn: HIRFunction,
+  sidemap: TemporariesSidemap,
+): Map<BlockId, BlockInfo> {
+  /**
+   * Due to current limitations of mutable range inference, there are edge cases in
+   * which we infer known-immutable values (e.g. props or hook params) to have a
+   * mutable range and scope.
+   * (see `destructure-array-declaration-to-context-var` fixture)
+   * We track known immutable identifiers to reduce regressions (as PropagateScopeDeps
+   * is being rewritten to HIR).
+   */
+  const knownImmutableIdentifiers = new Set<Identifier>();
+  if (fn.fnType === 'Component' || fn.fnType === 'Hook') {
+    for (const p of fn.params) {
+      if (p.kind === 'Identifier') {
+        knownImmutableIdentifiers.add(p.identifier);
+      }
+    }
+  }
+  const tree = new Tree();
+  const nodes = new Map<BlockId, BlockInfo>();
+  for (const [_, block] of fn.body.blocks) {
+    const propertyLoadInfos = sidemap.propertyLoadInfo.get(block.id);
+    CompilerError.invariant(propertyLoadInfos != null, {
+      reason: '[CollectHoistablePropertyLoads] block info missing',
+      loc: GeneratedSource,
+    });
+
+    const assumedNonNullObjects = new Set<PropertyLoadNode>();
+    for (const {instr, property} of propertyLoadInfos) {
+      const object = instr.value.object.identifier;
+      const propertyNode = tree.getPropertyLoadNode(property);
+      /**
+       * Since this runs *after* buildReactiveScopeTerminals, identifier mutable ranges
+       * are not valid with respect to current instruction id numbering.
+       * We use attached reactive scope ranges as a proxy for mutable range, but this
+       * is an overestimate as (1) scope ranges merge and align to form valid program
+       * blocks and (2) passes like MemoizeFbtAndMacroOperands may assign scopes to
+       * non-mutable identifiers.
+       *
+       * See comment at top of function for why we track known immutable identifiers.
+       */
+      const isMutableAtInstr =
+        object.mutableRange.end > object.mutableRange.start + 1 &&
+        object.scope != null &&
+        inRange(instr, object.scope.range);
```

Comment:
This is a little awkward, and also an example of being able to make use of more granular mutable range (instead of scope-range info)

Note that mutableRanges are no longer valid at this point, but we're only comparing relative start/ends (end === start + 1 indicates nonmutable value)